packages/rum-core/src/domain/rumEventsCollection/view/trackInteractionToNextPaint.ts

```diff
@@ -0,0 +1,113 @@
+import { type Duration, noop, isExperimentalFeatureEnabled, ExperimentalFeature } from '@datadog/browser-core'
+import { supportPerformanceTimingEvent, type RumEventTiming } from '../../../browser/performanceCollection'
+import { LifeCycleEventType, type LifeCycle } from '../../lifeCycle'
+import { ViewLoadingType } from '../../../rawRumEvent.types'
+import { getInteractionCount, initInteractionCountPolyfill } from './interactionCountPolyfill'
+
+// Arbitrary value to prevent unnecessary memory usage on views with lots of interactions.
+const MAX_INTERACTION_ENTRIES = 10
+
+/**
+ * Track the interaction to next paint (INP).
+ * To avoid outliers, return the p98 worst interaction of the view.
+ * Documentation: https://web.dev/inp/
+ * Reference implementation: https://github.com/GoogleChrome/web-vitals/blob/main/src/onINP.ts
+ */
+export function trackInteractionToNextPaint(viewLoadingType: ViewLoadingType, lifeCycle: LifeCycle) {
+  if (
+    !isInteractionToNextPaintSupported() ||
+    !isExperimentalFeatureEnabled(ExperimentalFeature.INTERACTION_TO_NEXT_PAINT)
+  ) {
+    return {
+      getInteractionToNextPaint: () => undefined,
+      stop: noop,
+    }
+  }
+
+  const { getViewInteractionCount } = trackViewInteractionCount(viewLoadingType)
+  const longestInteractions = trackLongestInteractions(getViewInteractionCount)
+  let maxInpDuration = -1 as Duration
+
+  const { unsubscribe: stop } = lifeCycle.subscribe(LifeCycleEventType.PERFORMANCE_ENTRIES_COLLECTED, (entries) => {
+    for (const entry of entries) {
+      if (entry.entryType === 'event' && entry.interactionId) {
+        longestInteractions.process(entry)
+      }
+    }
+
+    const inp = longestInteractions.estimateP98Duration()
+    if (inp && inp.duration > maxInpDuration) {
+      maxInpDuration = inp.duration
+    }
+  })
+
+  return {
+    getInteractionToNextPaint: () => {
+      // If no INP duration where captured because of the performanceObserver 40ms threshold
+      // but the view interaction count > 0 then report 0
+      if (maxInpDuration === -1 && getViewInteractionCount()) {
+        return 0 as Duration
+      } else if (maxInpDuration >= 0) {
+        return maxInpDuration
+      }
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏could be simplified a bit:
```suggestion
      if (maxInpDuration >= 0) {
        return maxInpDuration
      } else if (getViewInteractionCount()) {
        return 0 as Duration
      }
```